sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -128,17 +128,31 @@ abstract class BinaryArithmetic extends BinaryOperator with NullIntolerant {
   def calendarIntervalMethod: String =
     sys.error("BinaryArithmetics must override either calendarIntervalMethod or genCode")
 
+  def checkOverflowCode(result: String, op1: String, op2: String): String =
+    sys.error("BinaryArithmetics must override either checkOverflowCode or genCode")
+
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = dataType match {
     case _: DecimalType =>
       defineCodeGen(ctx, ev, (eval1, eval2) => s"$eval1.$decimalMethod($eval2)")
     case CalendarIntervalType =>
       defineCodeGen(ctx, ev, (eval1, eval2) => s"$eval1.$calendarIntervalMethod($eval2)")
+    // In the following cases, overflow can happen, so we need to check the result is valid.
+    // Otherwise we throw an ArithmeticException
```

Comment:
In current Spark we are very conservative about runtime error, as it may break the data pipeline middle away, and returning null is a commonly used strategy. Shall we follow it here? We can throw exception when we have a strict mode.

Reply:
Personally, I am quite against returning null. It is not something a user expects, so he/she is likely not to check for it (when I see a NULL myself, I think that one of the 2 operands was NULL, not that an overflow occurred), so he/she won't realize the issue and would find corrupted data. Moreover, this is not how RDBMS behaves and it is against SQL standard. So I think that the behavior which was chosen for DECIMAL was wrong and I'd prefer not to introduce the same behavior also in other places.

Anyway I see your point about consistency over the codebase and it makes sense.

I'd love to know @gatorsmile and @hvanhovell's opinions too.

Reply:
@gatorsmile @hvanhovell do you have time to check this and give your opinion here? Thanks.

Reply:
Why don't we fix it to return `null` and throw an exception when the configuration is on? Overflowed value is already quite pointless and changing the behaviour to return `null` by default might not be so harmful.

Reply:
Thanks for your comment @HyukjinKwon . The issue with returning null is described in https://github.com/apache/spark/pull/21599#issuecomment-409201709 (moreover this behavior would be against SQL standard, but that's a minor point).

Reply:
As I am working on https://github.com/apache/spark/pull/25300, +1 for returning null on overflow by default with @cloud-fan @HyukjinKwon . This makes arithmetic operations and casting consistent.
I think null is better than a non-sense number on overflow.

Reply:
I don't agree on this proposal for the reasons explained in the comment I mentioned earlier. Making all arithmetic operations nullable is a too broad change and I think it is not worth for the little benefit.